daemon/api_themes.go

```diff
@@ -0,0 +1,247 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	themesCmd = &Command{
+		Path:   "/v2/themes",
+		UserOK: true,
+		GET:    checkThemes,
+		POST:   installThemes,
+	}
+)
+
+type themeStatus string
+
+const (
+	themeInstalled   themeStatus = "installed"
+	themeAvailable   themeStatus = "available"
+	themeUnavailable themeStatus = "unavailable"
+)
+
+type themeStatusResponse struct {
+	GtkThemes   map[string]themeStatus `json:"gtk-themes"`
+	IconThemes  map[string]themeStatus `json:"icon-themes"`
+	SoundThemes map[string]themeStatus `json:"sound-themes"`
+}
+
+func getInstalledThemes(d *Daemon) (gtkThemes, iconThemes, soundThemes []string, err error) {
+	infos := d.overlord.InterfaceManager().Repository().Info(&interfaces.InfoOptions{
+		Names: []string{"content"},
+		Slots: true,
+	})
+	for _, info := range infos {
+		for _, slot := range info.Slots {
+			var content string
+			// The content interface ensures this attribute exists
+			if err := slot.Attr("content", &content); err != nil {
+				return nil, nil, nil, err
+			}
+			var themes *[]string
+			switch content {
+			case "gtk-3-themes":
+				themes = &gtkThemes
+			case "icon-themes":
+				themes = &iconThemes
+			case "sound-themes":
+				themes = &soundThemes
+			default:
+				continue
+			}
+			var sources []interface{}
+			if err := slot.Attr("source.read", &sources); err != nil {
+				continue
+			}
+			for _, s := range sources {
+				if path, ok := s.(string); ok {
+					*themes = append(*themes, filepath.Base(path))
+				}
+			}
+		}
+	}
+	sort.Strings(gtkThemes)
+	sort.Strings(iconThemes)
+	sort.Strings(soundThemes)
+	return gtkThemes, iconThemes, soundThemes, nil
+}
+
+var badPkgCharRegexp = regexp.MustCompile(`[^a-z0-9]+`)
+
+func themePackageCandidates(prefix, themeName string) []string {
```

Comment:
how many candidates will this produce on average?

Reply:
In the gtk-common-themes snap, we've got:

* 11 with one component
* 17 with two dash separated components
* 6 with three components.

Of the 21 multi-component theme names, 2 would need to be treated as complete on their own (Ambiant-MATE and Radiant-MATE), 16 would find a match by stripping off one component, and three from stripping two components.

So in the worst case of someone using a variant theme and it not being available in the store, it might generate three results.  If a theme snap is available, it's likely we'll hit a match after checking one or two candidates.

Of course, the outcome for the majority of users is that we never make it this far because their themes are already provided by the installed gtk-common-themes package.